plugins/ingest/src/main/java/org/elasticsearch/ingest/processor/HandledProcessor.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+
+package org.elasticsearch.ingest.processor;
+
+import org.elasticsearch.ingest.IngestDocument;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * A processor implementation may modify the data belonging to a document.
+ * Whether changes are made and what exactly is modified is up to the implementation.
+ */
+public class HandledProcessor implements Processor {
```

Comment:
I think that embedding the on failure handling here is good. I've been thinking about how to make this class also usable on the pipeline level:

What I think we can change here is to let `HandledProcessor` also handle multiple processors, so that then this class can also be used at the pipeline level.

Something like this:
```java
public class HandledProcessor implements Processor {

    // No need to use HandledProcessor as generic type, since that is an implementation detail
    private final List<Processor> processors; 
    private final List<Processor> onFailureProcessors;

    public HandledProcessor(Processor... processor) {
        this(Arrays.asList(processor), Collections.emptyList());
    }
    public HandledProcessor(List<Processor> processors, List<Processor> onFailureProcessors) {
        this.processors = Collections.unmodifiableList(processors);
        this.onFailureProcessors = Collections.unmodifiableList(onFailureProcessors);
    }

    public boolean isHandled() {
        return !onFailureProcessors.isEmpty();
    }

    public List<Processor> getProcessors() {
        return processors;
    }

    @Override
    public void execute(IngestDocument ingestDocument) throws Exception {
        try {
            for (Processor processor : processors) {
                processor.execute(ingestDocument);
            }
        } catch (Exception e) {
            if (isHandled()) {
                executeOnFailure(ingestDocument);
            } else {
                throw e;
            }
        }

    }

    public void executeOnFailure(IngestDocument ingestDocument) throws Exception {
        for (Processor processor : onFailureProcessors) {
            processor.execute(ingestDocument);
        }
    }
}
```

Then in `Pipeline` the `processors` field type can be changed to `HandledProcessor`.
The the pipeline factory can then benefit from having the same parsing logic for the on_failure on both processor and pipeline level. 

However the `Processor#getType()` method is problematic with this approach, because it doesn't assume a processor is embedding multiple processors. This method is only really used in the simulate api and I think for that reason this method needs to be removed from the `Processor` interface and the functionality of figuring out what is the processor type should go somewhere else maybe the `PipelineStore`? We can do this in a followup pr and for now just throw an `UnsupportedOperationException` in HandledProcessor#getType(). The `Pipeline#getProcessors()` method then just needs to delegate to  `HandledProcessor#getProcessors` so that the `SimulateExecutionService` doesn't fail.

If this makes sense, then we should rename `HandledProcessor` to `CompoundProcessor`.